### PR TITLE
feat(static-module-record): add support for import.meta

### PR DESCRIPTION
**Pull request by naugtur**
Implemented import.meta according to #291

The actual implementation of exposing import.meta.url to modules loaded by compartment-mapper is going to be a separate PR.

### CODE REVIEW

**Review by kriskowal — COMMENTED** [review]
Let’s talk requirements for this feature at Endo Sync tomorrow. The Compartment abstraction will be aware of meta but not meta.url. For that, we need to give the hooks system a way to populate meta. The current notion is that the import/loadHook will be able to return an object with a meta property and the compartment will copy the properties to the meta object. Because some of those properties are expensive, we would have an importMetaHook that takes a meta object and add things like resolve but would only be called if a module utters import.meta in its source text.

**Comment by kriskowal** [comment]
See https://github.com/endojs/endo/issues/291

File: packages/compartment-mapper/src/bundle.js
Comment (naugtur) [inline packages/compartment-mapper/src/bundle.js:337]: What do we want to do about bundler?

**Comment by naugtur** [comment]
Made some progress on the implementation. Looks like it's it.

File: packages/compartment-mapper/src/bundle.js
Comment (kriskowal) [reply packages/compartment-mapper/src/bundle.js:337]: To support `import.meta` in the bundler, we’d inline the logic we’re introducing to `module-instance.js` in the generated code. `Object.assign($h_ort_meta, descriptor.meta)`. We’d need to capture the `meta` property and JSON serialize it into the bundle at that location.

For now, though, the bundler is very incomplete and can remain that way until we have a motivating use case, like replacing Browserify at MetaMask or replacing Rollup in Agoric SDK bootstrap. To get there, it needs a lot of work. For now, it’s sufficient for generating the SES bundles only. For example, it only works within a single package. It doesn’t support CommonJS (though, it probably could pretty easily).

**Review by kriskowal — COMMENTED** [review]
Yeah, this is on-track. The comments I’ve left pertain to the leading half of this feature, which needs to allow for usage like:

```js
const c = new Compartment({}, {}, {
  importHook(specifier) {
    const requestUrl = locate(specifier);
    const response = await fetch(requestUrl);
    const responseText = await response.text();
    return {
      record: new StaticModuleRecord(responseText),
      meta: {url: responseURL};
    };
  },
  importMetaHook(meta) {
    import.meta.resolve = url => importMetaResolve(url, meta.url);
  },
});
```

File: packages/ses/src/module-instance.js
Comment (naugtur) [reply packages/ses/src/module-instance.js:166]: Ok, that looks very similar to what we saw from Modable too.

File: packages/compartment-mapper/src/bundle.js
Comment (naugtur) [reply packages/compartment-mapper/src/bundle.js:335]: As a newcomer to the codebase I enjoyed the variables all sharing a naming convention. It was much easier to spot them in the more complex bits. Could we incorporate the term "a value for a hidden variable" into the name somehow?

File: packages/ses/src/module-instance.js
Comment (naugtur) [reply packages/ses/src/module-instance.js]: ok, keeping it room temperature for now.

File: packages/compartment-mapper/src/bundle.js
Comment (naugtur) [reply packages/compartment-mapper/src/bundle.js:337]: this is now potentially the only remaining todo unless we want to progress on changing the record envelope format

File: packages/ses/src/module-load.js
Comment (naugtur) [reply packages/ses/src/module-load.js:324]: my bad. I always check if I committed it before the final merge, I change it so often. Thinking of automating it with a precommit script

File: packages/static-module-record/src/static-module-record.js
Comment (naugtur) [reply packages/static-module-record/src/static-module-record.js:66]: ok. I might have fixated on the Node.js `resolve` being expensive.

File: packages/ses/README.md
Comment (naugtur) [inline packages/ses/README.md:422]: I suppose we're going to need to write more about importMetaHook

File: packages/static-module-record/src/transform-analyze.js
Comment (naugtur) [inline packages/static-module-record/src/transform-analyze.js:68]: I don't like the looks of it, but how else do I pass a boolean by reference? 
At least it's future proof if we want more information out of it.

File: packages/compartment-mapper/src/bundle.js
Comment (naugtur) [reply packages/compartment-mapper/src/bundle.js:337]: So should we agree on how to proceed for bundle.js or should I remove this change and add a new issue?

File: packages/static-module-record/test/test-static-module-record.js
Comment (naugtur) [reply packages/static-module-record/test/test-static-module-record.js:563]: Update.

`rewriteDeclaration` is somehow preventing visitor from reaching its results. 
AFAIK visitor should go into replacements and run on nodes matched there. I don't know why this is not happening here. 

Adding `path.traverse(...` inside `rewriteDeclaration` can fix the issue, but it halves the results of a quick benchmark I've set up.

One way to keep the performance unchanged is to rewrite on the first pass not the second. It works, but it's probably controversial.

Comment (naugtur) [reply packages/static-module-record/test/test-static-module-record.js:563]: I've got n update where `rewriteDeclarations` no longer escapes the visitor. I managed to narrow it down to something that replacing declarations with themselves was doing. It seems ok to append instead.

I'll make a separate PR with it.

Comment (naugtur) [reply packages/static-module-record/test/test-static-module-record.js]: separate PR merged

File: packages/compartment-mapper/src/bundle.js
Comment (naugtur) [reply packages/compartment-mapper/src/bundle.js:337]: Agreed to not do anything and leave import.meta as an empty object.

**Comment by naugtur** [comment]
@kriskowal Unless I need to put more contrnt into `packages/ses/README.md` I think this is ready for final review.

File: packages/ses/README.md
Comment (kriskowal) [reply packages/ses/README.md:422]: We probably also ought to make the signature `importMetaHook(fullSpecifier, importMeta)` in case that’s useful.

File: packages/ses/src/module-instance.js
Comment (kriskowal) [inline packages/ses/src/module-instance.js:173]: We probably won’t regret making this `importMetaHook(moduleSpecifier, importMeta)`

File: packages/static-module-record/src/babelPlugin.js
Comment (kriskowal) [reply packages/static-module-record/src/babelPlugin.js:302]: I like the specificity.

File: packages/static-module-record/src/hidden.js
Comment (kriskowal) [inline packages/static-module-record/src/hidden.js:8]: One of us requested that we somehow make the apparent width of the replacement string the same as the original, so it doesn’t displace the column number of following text. I’m not sure how ZWJ generally appears in, say, VSCode, but maybe this should be something like`$h_____meta`.

File: packages/static-module-record/src/static-module-record.js
Comment (kriskowal) [inline packages/static-module-record/src/static-module-record.js:64]: The tentative proposed verbiage is “needs”. This is “pretend not to see” public. Which is to say, we don’t get to change it after it releases. So perhaps we should wait for a blessing on https://github.com/tc39/proposal-compartments/pull/46/ before we land this.

File: packages/static-module-record/src/transform-analyze.js
Comment (kriskowal) [reply packages/static-module-record/src/transform-analyze.js:29]: Passing an `importMetaSeen()` function instead might be clearer.

File: packages/static-module-record/src/transform-analyze.js
Comment (kriskowal) [inline packages/static-module-record/src/transform-analyze.js:93]: I believe we agree that we should pretend “utter” isn’t a word.

**Review by kriskowal — COMMENTED** [review]
This is ready to land, though the specification it’s based on needs a little more time to bake. Until then, I’ve left some notes for your consideration.

**Review by kriskowal — APPROVED** [review]
…and I’m leaving a stamp so you don’t have to wait for my blessing when everything’s ready.

File: packages/static-module-record/src/hidden.js
Comment (naugtur) [reply packages/static-module-record/src/hidden.js:8]: ZWJ can display as nothing or as a blocky box and a warning that  something's fishy about the file. I guess it's the number of characters that counts, so ZWJ counts as 1.

I'll adjust the length and add a comment explaining what happened.

File: packages/static-module-record/src/transform-analyze.js
Comment (naugtur) [reply packages/static-module-record/src/transform-analyze.js:93]: That was 2 weeks after I put it here based on the sophisticated-vocabulary-filled meeting ;)
Then again - this is an internal name visible to people who need to grow their vocabulary significantly to grok this codebase anyway.  
I think the best result would be to make it consistent with the public name we eventually land.
https://github.com/tc39/proposal-compartments/pull/46

File: packages/ses/README.md
Comment (naugtur) [reply packages/ses/README.md]: done

File: packages/static-module-record/src/hidden.js
Comment (naugtur) [reply packages/static-module-record/src/hidden.js]: length including ZWJj should now match

**Pull request merged**
